### ISSUE TITLE
Different coco eval metrics right after training and when loaded from pth file

### ISSUE THREAD

**Issue by shridharmishra4** [facebookresearch/detectron2#1894]
## Instructions To Reproduce the Issue:

Followed below tutorial to fine tune the model
https://www.dlology.com/blog/how-to-train-detectron2-with-custom-coco-datasets/

Simplify the steps to reproduce the issue using suggestions from the above link, and provide them below:

**Training code:**
```
from detectron2.engine import DefaultTrainer
from detectron2.config import get_cfg
import os
from detectron2 import model_zoo


cfg = get_cfg()
# cfg.merge_from_file("../detectron2/configs/COCO-Detection/faster_rcnn_R_50_FPN_1x.yaml")
# cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-Detection/faster_rcnn_R_50_FPN_1x.yaml")

cfg.merge_from_file("../detectron2/configs/COCO-Detection/faster_rcnn_R_101_FPN_3x.yaml")
cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-Detection/faster_rcnn_R_101_FPN_3x.yaml")



cfg.DATASETS.TRAIN = ("Hands",)
cfg.DATASETS.TEST = ()   # no metrics implemented for this dataset
# cfg.TEST.EVAL_PERIOD = 10

cfg.DATALOADER.NUM_WORKERS = 4

cfg.SOLVER.IMS_PER_BATCH = 16
cfg.SOLVER.BASE_LR = 0.00125
cfg.SOLVER.STEPS = (120000,160000)
cfg.SOLVER.MAX_ITER = 4000    # 300 iterations seems good enough, but you can certainly train longer
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 128   # faster, and good enough for this toy dataset
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 2  # 2 classes (Hand_Left, Hand_Right)
cfg.MODEL.BACKBONE.FREEZE_AT = 2
print(cfg)

now = datetime.now()
date_time = now.strftime("%d_%m_%Y-%H:%M:%S")
models_dir = "./models/v-{}-timestamp-{}".format(version,date_time)

cfg.OUTPUT_DIR = models_dir
os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg)
trainer.resume_or_load(resume=False)
trainer.train()

```
**Evaluation after training**

```
from detectron2.evaluation import COCOEvaluator, inference_on_dataset
from detectron2.data import build_detection_test_loader

cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.85   # set the testing threshold for this model
cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST = 0.7
cfg.DATASETS.TEST = ("HandsValidation", )
predictor = DefaultPredictor(cfg)
evaluator = COCOEvaluator("HandsValidation", cfg,
                          False, output_dir=cfg.OUTPUT_DIR)

val_loader = build_detection_test_loader(cfg, "HandsValidation")
inference_on_dataset(trainer.model, val_loader, evaluator)
```

**Evaluation from disk:**
```
from detectron2.evaluation import COCOEvaluator, inference_on_dataset
from detectron2.data import build_detection_test_loader
from detectron2.config import get_cfg
from detectron2.utils.visualizer import ColorMode
from detectron2.engine import DefaultPredictor
import random
import detectron2
from detectron2.utils.logger import setup_logger
setup_logger()

# import some common libraries
import matplotlib.pyplot as plt
import numpy as np
import cv2
import os
print(np.__version__)
os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # Or 2, 3, etc. other than 0


# import some common detectron2 utilities
from detectron2.config import get_cfg
from detectron2.utils.visualizer import Visualizer
from detectron2.data import MetadataCatalog, DatasetCatalog

import os

cfg = get_cfg()

cfg.merge_from_file("../detectron2/configs/COCO-Detection/faster_rcnn_R_101_FPN_3x.yaml")
cfg.MODEL.WEIGHTS = os.path.join("./models/v-0.05-timestamp-14_07_2020-19:51:36", "model_final.pth")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.85   # set the testing threshold for this model
cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST = 0.7
cfg.DATASETS.TEST = ("HandsValidation", )
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 2  # 2 classes (Hand_Left, Hand_Right)

predictor = DefaultPredictor(cfg)
idx = 0
for d in random.sample(validation_dataset_dicts, 5):
    idx+=1
    im = cv2.imread(d["file_name"])
    outputs = predictor(im)
    v = Visualizer(im[:, :, ::-1],
                   metadata=hands_validation_meta_data, 
                   scale=0.85, 
                   
    )
    vis = v = v.draw_instance_predictions(outputs["instances"].to("cpu"))
    plt.figure()
#     plt.imshow(vis.get_image()[:, :, ::-1])
    plt.axis('off')
    plt.imshow(cv2.cvtColor(vis.get_image()[:, :, ::-1], cv2.COLOR_BGR2RGB))
    plt.axis('off')
    plt.savefig("./sample_images/{}.png".format(idx),dpi=300)

evaluator = COCOEvaluator("HandsValidation", cfg,
                          False, output_dir=cfg.OUTPUT_DIR)

val_loader = build_detection_test_loader(cfg, "HandsValidation")
inference_on_dataset(predictor.model, val_loader, evaluator)
```

## Expected behavior:

COCO evaluation right after training achieves a higher accuracy than reloading the the same model and then running eval.
Its something like

**Right after training**
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.459
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.852
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.450
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.417
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.506
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.546
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.580
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.580
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.527
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.630

**Loading the same model after saving it.**
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.417
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.768
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.420
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.380
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.460
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.502
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.515
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.515
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.467
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.561
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = -1.000

Any ideas on what might be going wrong? 

## Environment:

```
which: no hipcc in (/nfs/hpc/share//miniconda/envs/hand/bin:/nfs/hpc/share//anaconda/bin:/usr/local/apps/cudnn/cudnn-10.0/lib64:/apps/slurm/20.02/bin:/usr/local/cuda/bin:/opt/bin:/nfs/hpc/share/miniconda/bin:/nfs/hpc/share//miniconda/condabin:/nfs/hpc/share//anaconda/bin:/usr/local/apps/cudnn/cudnn-10.0/lib64:/usr/lib64/qt-3.3/bin:/nfs/stak/users//perl5/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/usr/local/apps/bin:/nfs/stak/users//bin:/scratch/a1/Univa-8.6.4/bin/lx-amd64:/opt/ibutils/bin)
------------------------  ------------------------------------------------------------------------------------------------
sys.platform              linux
Python                    3.6.10 |Anaconda, Inc.| (default, May  8 2020, 02:54:21) [GCC 7.3.0]
numpy                     1.17.4
detectron2                0.1.3 
detectron2 compiler       GCC 7.3
detectron2 CUDA compiler  10.1
detectron2 arch flags     sm_35, sm_37, sm_50, sm_52, sm_60, sm_61, sm_70, sm_75
DETECTRON2_ENV_MODULE     <not set>
PyTorch                   1.5.0+cu101 
PyTorch debug build       False
CUDA available            True
GPU 0                     Tesla V100-SXM3-32GB
CUDA_HOME                 /usr/local/apps/cuda/cuda-10.0/
NVCC                      Cuda compilation tools, release 10.0, V10.0.130
Pillow                    6.2.2
torchvision               0.6.0+cu101 
torchvision arch flags    sm_35, sm_50, sm_60, sm_70, sm_75
fvcore                    0.1.1
cv2                       4.2.0
------------------------  ------------------------------------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2019.0.5 Product Build 20190808 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.21.1 (Git Hash 7d2fd500bc78936d1d648ca713b901012f470dbc)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.2
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DNDEBUG -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 
```

**Comment by ppwwyyxx** [comment]
> Evaluation after training
from detectron2.evaluation import COCOEvaluator, inference_on_dataset
from detectron2.data import build_detection_test_loader
cfg.MODEL.WEIGHTS = os.path.join(cfg.OUTPUT_DIR, "model_final.pth")
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.85   # set the testing threshold for this model
cfg.MODEL.ROI_HEADS.NMS_THRESH_TEST = 0.7
cfg.DATASETS.TEST = ("HandsValidation", )
predictor = DefaultPredictor(cfg)
evaluator = COCOEvaluator("HandsValidation", cfg,
                          False, output_dir=cfg.OUTPUT_DIR)
val_loader = build_detection_test_loader(cfg, "HandsValidation")
inference_on_dataset(trainer.model, val_loader, evaluator)

here you evaluated `trainer.model` which was constructed using the config without setting `SCORE_THRESH_TEST` or `NMS_THRESH_TEST`
